network/src/protocols/wire/messaging/v1/mod.rs

```diff
@@ -154,27 +159,58 @@ pub fn network_message_frame_codec(max_frame_size: usize) -> LengthDelimitedCode
 /// underlying socket.
 #[pin_project]
 pub struct NetworkMessageStream<TReadSocket: AsyncRead> {
+    ip_addr: IpAddr,
     #[pin]
     framed_read: FramedRead<IoCompat<TReadSocket>, LengthDelimitedCodec>,
+    rate_limiter: Option<Arc<RateLimiter<IpAddr>>>,
 }
 
 impl<TReadSocket: AsyncRead> NetworkMessageStream<TReadSocket> {
     pub fn new(socket: TReadSocket, max_frame_size: usize) -> Self {
+        Self::new_with_rate_limiting(
+            IpAddr::V4(Ipv4Addr::UNSPECIFIED),
+            socket,
+            max_frame_size,
+            None,
+        )
+    }
+
+    pub fn new_with_rate_limiting(
+        ip_addr: IpAddr,
+        socket: TReadSocket,
+        max_frame_size: usize,
+        rate_limiter: Option<Arc<RateLimiter<IpAddr>>>,
+    ) -> Self {
         let frame_codec = network_message_frame_codec(max_frame_size);
         let compat_socket = IoCompat::new(socket);
         let framed_read = FramedRead::new(compat_socket, frame_codec);
-        Self { framed_read }
+        Self {
+            ip_addr,
+            framed_read,
+            rate_limiter,
+        }
     }
 }
 
 impl<TReadSocket: AsyncRead> Stream for NetworkMessageStream<TReadSocket> {
     type Item = Result<NetworkMessage, ReadError>;
 
     fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
-        match self.project().framed_read.poll_next(cx) {
+        let stream = self.project();
+        match stream.framed_read.poll_next(cx) {
             Poll::Ready(Some(Ok(frame))) => {
                 let frame = frame.freeze();
 
+                // Handle rate limiting of messages
+                if let Some(rate_limiter) = &stream.rate_limiter {
+                    if let Err(_error) = rate_limit_msg(rate_limiter, &stream.ip_addr, frame.len())
+                    {
+                        // TODO save message until it's ready
+                        // Drop message, it's been throttled
+                        return Poll::Pending;
```

Comment:
the `Stream` will hang here if we just return `Poll::Pending` without registering our `cx.waker` somewhere. When we return `Poll::Pending`, we're telling the parent future/stream/runtime: "hey, I don't have any more work to do and you don't need to poll me again until someone has triggered my waker`. 

This is why we pass our `cx` (which contains our waker) to our child `framed_read.poll_next(cx)`. When there's no more data on the socket, it'll return `Poll::Pending` and register the waker with the epoll reactor to wake us up when there's more data available.

In this case, we probably want to sleep a bit or maybe just yield immediately back to the scheduler like

```rust
cx.waker.wake();
return Poll::Pending;
```

Sleeping is tougher:

```rust
use futures::future::Future;
use futures::ready;
use std::time::Duration;
use tokio::time::{delay_for, Delay};

#[pin_project]
pub struct NetworkMessageStream {
    #[pin]
    framed_read: /* .. */,
    // When we have Some(delay), we'll always finish polling it (the delay has finished)
    // before polling the framed_read.
    // Note: don't need pin here b/c Delay: Unpin
    delay: Option<Delay>,
}

impl NetworkMessageStream {
    pub fn new(socket, max_frame_size) -> Self {
        // ..
        Self { framed_read, delay: None }
    }
}

impl Stream for NetworkMessageStream {
    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        if self.delay.is_some() {
            // ready! will return Poll::Pending if delay.poll(cx) returns Poll::Pending
            ready!(Pin::new(self.as_mut().project().delay.as_mut().unwrap()).poll(cx));

            // delay is done, set it to None
            *self.as_mut().project().delay = None;
        }

        // finally read the next frame
        match self.as_mut().project().framed_read.poll_next(cx) {
            Poll::Ready(Some(Ok(frame))) => {
                let frame = frame.freeze();

                // we want to delay before reading the next frame, so we set delay to Some(Delay)
                if rate_limited {
                    let delay_time = /* .. */;
                    *self.as_mut().project().delay = Some(delay_for(Duration::from_millis(delay_time)));
                }
                
                // ..
        }
    }

```